omnigibson/transition_rules.py

```diff
@@ -103,551 +263,952 @@ def __call__(self, obj):
+                # Otherwise, all conditions met, we found a valid recipe and so we execute and terminate early
+                og.log.info(f"Executing recipe: {name} in container {container.name}!")
+
+                # Take the transform
+                t_results.remove += cls._transform_to_output_system(
+                    container=container,
+                    output_system=recipe["output_system"],
+                    in_volume=in_volume,
+                )
+
+                # Terminate early
+                return t_results
+
+            # Otherwise, if we didn't find a valid recipe, we execute a garbage transition instead
+            og.log.info(f"Did not find a valid recipe; generating {m.DEFAULT_GARBAGE_SYSTEM} in {container.name}!")
+
+            # Generate garbage fluid
+            t_results.remove += cls._transform_to_output_system(
+                container=container,
+                output_system=m.DEFAULT_GARBAGE_SYSTEM,
+                in_volume=in_volume,
+            )
+
+            return t_results
+
+    @classmethod
+    def _transform_to_output_system(cls, container, output_system, in_volume):
+        """
+        Transforms all items contained in @container into @output_system, generating volume of @output_system
+        proportional to the number of items transformed.
+
+        Args:
+            container (BaseObject): Container object which will have its contained elements transformed into
+                @output_system
+            output_system (str): Name of the output system to generate. Should be a PhysicalParticleSystem
+            in_volume (n-array): (n_objects,) boolean array specifying whether every object from og.sim.scene.objects
+                is contained in @container or not. Only necess
+
+        Returns:
+            list of BaseObject: List of object(s) to remove from the simulator due to this transformation
+        """
+        # Compute total volume of all contained items
+        volume = 0
+        objects_to_remove = []
+
+        # Remove all recipe system particles contained in the container
+        for system in PhysicalParticleSystem.get_active_systems():
+            if container.states[Contains].get_value(system):
+                volume += contained_particles_state.get_value()[0] * np.pi * (system.particle_radius ** 3) * 4 / 3
+                container.states[Filled].set_value(system, False)
+        for system in VisualParticleSystem.get_active_systems():
+            group_name = system.get_group_name(container)
+            if group_name in system.groups and system.num_group_particles(group_name) > 0:
+                system.remove_all_group_particles(group=group_name)
+
+        # Remove all contained objects if requested
+        if not cls.ignore_nonrecipe_objects:
+            # Remove all objects inside the container
+            objs_to_remove = cls._OBJECTS[np.where(in_volume)[0]].tolist()
+            volume += sum(obj.volume for obj in objs_to_remove)
```

Comment:
for mixing_ward,

if matching recipe: the recipe-relevant object should be removed, other rigid objects stay intact
if not (garbage): all rigid bodies should stay intact

Reply:
whoops! great point. fixed (tho to be clear: mixing_wand rules should never have any recipe-relevant objects, right?)